### PR TITLE
[GDPR 11] Promote minor account to adult account

### LINKED ISSUE #20: [GDPR 11] Promote minor account to adult account
Ref: GDPR 11

**Is your feature request related to a problem? Please describe.**
As a registered user with a minor account, I want to be able to promote my account when I have reached the legal age of consent for data protection purposes, to have an adult (regular) account.

**Describe the solution you'd like**
When the minor reaches the age of majority, an email will be sent to the minor and the guardian to let them know that their accounts are now independent.

This task will be executed automatically when the minor exceeds the age from which the guardian's authorisation is no longer required (ref: GDPR09).

**Describe alternatives you've considered**
Another possibility would be that the minor is the one who directly promotes his or her account to the adult.

This implies thinking about what happens if the minor does not do this promotion. This would require us to perform an automatic task once a year to automatically promote the accounts of minors who are over the maximum age.

As the access for minors depends on the context of each organisation, this development should be a module.

**Does this issue could impact on users private data?**
Yes as it afects to minors data, it is particulary sensitive.

**Acceptance criteria**

- [ ] As a registered minor when I reach the legal age of consent my account is automatically promoted to adult.
- [ ] As a registered minor when I reach the legal age of consent I receive an email stating that my account has been promoted to adult.

### PR DESCRIPTION
**Pull request by apuertadelaguila**
Fixes #20

With this PR, we can promote through a rake task, a minor account which has reached the maximum minor age of the organization, to an adult user account.

- [x] As a registered minor when I reach the legal age of consent, my account is automatically promoted to adult.
- [ ]  As a registered minor when I reach the legal age of consent, I receive an email stating that my account has been promoted to adult.

### CODE REVIEW

File: spec/lib/tasks/kids_spec.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe "kids:promote_minor_accounts", type: :task do
+  let(:organization) { create(:organization, available_authorizations: %w(dummy_authorization_handler)) }
+  let(:user) { create(:user, :admin, :confirmed, organization:) }
+  let(:other_user) { create(:user, :admin, :confirmed, organization:) }
+  let(:minor) { create(:minor, tutor: user, organization:) }
+  let(:minor_to_promote) { create(:minor_to_promote, tutor: user, organization:) }
```
Comment (microstudi) [inline spec/lib/tasks/kids_spec.rb]:
```suggestion
  let(:minor_to_promote) { create(:minor, tutor: user, minor_data_to_promote:, organization:) }
  let(:minor_data_to_promote) { create :minor_data, birthday: }
  let(:birthday) { 14.years.ago }
```
I found this extra factory confusing, can we just set the birthday instead?
This also would let you check several ages. 
At least we also should test the consecutive age when an account is promoted (from 13 to 14 yo)

File: app/models/decidim/kids/minor_account.rb [hunk 2]
```diff
@@ -16,6 +16,13 @@ class MinorAccount < ApplicationRecord
       validate :can_be_tutor
       validate :can_be_minor
 
+      def promote_account
+        return if minor.minor_age < (minor&.organization&.maximum_minor_age || Decidim::Kids.maximum_minor_age)
```
Comment (microstudi) [inline app/models/decidim/kids/minor_account.rb]:
```suggestion
        return if minor.minor_age <= (minor&.organization&.maximum_minor_age || Decidim::Kids.maximum_minor_age)
```
If the minor matches the maximum_minor_age is still a minor right?

File: lib/decidim/kids/test/factories.rb [hunk 3]
```diff
@@ -24,6 +24,17 @@
...
+    transient do
+      tutor { create(:user, :confirmed, organization:) }
+    end
+
+    after(:create) do |user, evaluator|
+      create(:minor_to_promote_account, tutor: evaluator.tutor, minor: user)
+      create(:minor_to_promote_data, user:)
+    end
+  end
+
```
Comment (microstudi) [inline lib/decidim/kids/test/factories.rb]: I found this unnecessary

File: lib/decidim/kids/test/factories.rb [hunk 4]
```diff
@@ -41,10 +52,22 @@
     minor { create(:user, :confirmed, organization: tutor.organization) }
   end
 
+  factory :minor_to_promote_account, class: "Decidim::Kids::MinorAccount" do
+    tutor { create(:user, :confirmed, organization:) }
+    minor { create(:user, :confirmed, organization: tutor.organization) }
+  end
+
```
Comment (microstudi) [inline lib/decidim/kids/test/factories.rb]: This too

File: lib/decidim/kids/test/factories.rb [hunk 5]
```diff
@@ -41,10 +52,22 @@
...
-    birthday { rand(10..14).years.ago }
+    birthday { rand(10..13).years.ago }
+  end
+
+  factory :minor_to_promote_data, class: "Decidim::Kids::MinorData" do
+    user { create(:minor_to_promote) }
+    name { Faker::Name.name }
+    email { Faker::Internet.email }
+    birthday { 15.years.ago }
   end
```
Comment (microstudi) [inline lib/decidim/kids/test/factories.rb]: And this

File: app/models/decidim/kids/minor_account.rb [hunk 6]
```diff
@@ -16,6 +16,13 @@ class MinorAccount < ApplicationRecord
       validate :can_be_tutor
       validate :can_be_minor
 
+      def promote_account
```
Comment (microstudi) [inline app/models/decidim/kids/minor_account.rb]:
```suggestion
      def promote_account!
```
As this is destructive, maybe this convention would be better

File: app/models/decidim/kids/minor_account.rb [hunk 7]
```diff
@@ -16,6 +16,13 @@ class MinorAccount < ApplicationRecord
...
+        return if minor.minor_age < (minor&.organization&.maximum_minor_age || Decidim::Kids.maximum_minor_age)
+
+        minor.tutors = []
```
Comment (microstudi) [inline app/models/decidim/kids/minor_account.rb]: In here we should make sure that we are destroying any associated MinorData too.

File: spec/lib/tasks/kids_spec.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    context "when removing minor account relation" do
+      it "remove no longer minor accounts relations" do
+        expect(organization.maximum_minor_age).to eq(13)
+        expect(minor_to_promote.minor_age).to eq(15)
+        task.execute
+        expect(Decidim::Kids::MinorAccount.where(minor: minor_to_promote)).to be_empty
+        expect(Decidim::Kids::MinorAccount.where(minor:)).not_to be_empty
+      end
+
```
Comment (microstudi) [inline spec/lib/tasks/kids_spec.rb:49]:
```suggestion
      it " to the new regular users" do
        expect { task.execute }.to change { Decidim::Kids::MinorAccount.count }.by(-1)
      end
```
A possible extra test

**Review by microstudi — CHANGES_REQUESTED** [review]
I've found a couple of issues, please see the comments.

A part from this, we should include in the README some instructions about recomending adding the Rake task in a cron system. Similarly as it is done in other places.

File: spec/lib/tasks/kids_spec.rb (hunk 1 above)
Comment (apuertadelaguila) [reply spec/lib/tasks/kids_spec.rb]: I've tried to replicate the suggestions. In this specific case, I wasn't able to create "minor_to_promote" by creating another "minor" passing a "birthday" in its "minor_data_to_promote". "minor_data_to_promote" creates a ":minor_data" and this one a new minor user who creates another ":minor_data" and overrides the given "birthday"

Comment (microstudi) [reply spec/lib/tasks/kids_spec.rb]: Right, the problem is that the factory `minor` is just a convenient way to create a "minor user". In your case,you just need to define all the properties of what makes a user a "minor":

```ruby
  let!(:minor_to_promote) { create(:user, :confirmed, organization:) }
  let!(:minor_to_promote_account) { create :minor_account, tutor: user, minor: minor_to_promote}
  let!(:minor_data_to_promote) { create :minor_data, user: minor_to_promote, birthday: }
  let(:birthday) { 15.years.ago }
```

Btw, once you do this, you can add a context and test the case for 14 years old.

Comment (apuertadelaguila) [reply spec/lib/tasks/kids_spec.rb]: Done

File: app/models/decidim/kids/minor_account.rb (hunk 2 above)
Comment (apuertadelaguila) [reply app/models/decidim/kids/minor_account.rb]: Done

File: lib/decidim/kids/test/factories.rb (hunk 3 above)
Comment (apuertadelaguila) [reply lib/decidim/kids/test/factories.rb]: Done

File: lib/decidim/kids/test/factories.rb (hunk 5 above)
Comment (apuertadelaguila) [reply lib/decidim/kids/test/factories.rb]: Done

File: app/models/decidim/kids/minor_account.rb (hunk 6 above)
Comment (apuertadelaguila) [reply app/models/decidim/kids/minor_account.rb]: Done

File: spec/lib/tasks/kids_spec.rb (hunk 8 above)
Comment (apuertadelaguila) [reply spec/lib/tasks/kids_spec.rb:49]: Done

File: app/models/decidim/kids/minor_account.rb (hunk 7 above)
Comment (apuertadelaguila) [reply app/models/decidim/kids/minor_account.rb]: Done

File: lib/decidim/kids/test/factories.rb (hunk 4 above)
Comment (apuertadelaguila) [reply lib/decidim/kids/test/factories.rb]: Done

File: spec/lib/tasks/kids_spec.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  let!(:minors_organization_config) { create(:minors_organization_config, organization:, enable_minors_participation:) }
+
+  before do
+    minor.confirm
+  end
+
+  context "when minors turn the maximum minor age" do
+    it "run gracefully" do
+      expect(Decidim::Kids::MinorAccount.where(minor:)).to be_present
+      expect(Decidim::Kids::MinorAccount.where(minor: minor_to_promote)).to be_present
```
Comment (microstudi) [inline spec/lib/tasks/kids_spec.rb]:
```suggestion
      expect(Decidim::Kids::MinorAccount.where(minor:)).to be_present
      expect(Decidim::Kids::MinorAccount.where(minor: minor_to_promote)).to be_present
      expect(Decidim::Kids::MinorData.where(user: minor)).to be_present
      expect(Decidim::Kids::MinorData.where(user: minor_to_promote)).to be_present
```

File: spec/lib/tasks/kids_spec.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+      expect { task.execute }.not_to raise_error
+    end
+
+    context "when removing minor account relation" do
+      it "removes no longer minor accounts relations" do
+        expect(organization.maximum_minor_age).to eq(13)
+        expect(minor_to_promote.minor_age).to eq(15)
+        task.execute
+        expect(Decidim::Kids::MinorAccount.where(minor: minor_to_promote)).to be_empty
+        expect(Decidim::Kids::MinorAccount.where(minor:)).not_to be_empty
```
Comment (microstudi) [inline spec/lib/tasks/kids_spec.rb]:
```suggestion
        expect(Decidim::Kids::MinorAccount.where(minor: minor_to_promote)).to be_empty
        expect(Decidim::Kids::MinorAccount.where(minor:)).not_to be_empty
        expect(Decidim::Kids::MinorData.where(user: minor_to_promote)).to be_empty
        expect(Decidim::Kids::MinorData.where(user: minor)).not_to be_empty
```

File: spec/lib/tasks/kids_spec.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    it "run gracefully" do
+      expect(Decidim::Kids::MinorAccount.where(minor:)).to be_present
+      expect(Decidim::Kids::MinorAccount.where(minor: minor_to_promote)).to be_present
+      expect { task.execute }.not_to raise_error
+    end
+
+    context "when removing minor account relation" do
+      it "removes no longer minor accounts relations" do
+        expect(organization.maximum_minor_age).to eq(13)
+        expect(minor_to_promote.minor_age).to eq(15)
```
Comment (microstudi) [inline spec/lib/tasks/kids_spec.rb]: I think we should add 2 tests cases more:

1. when the user is exaclty one year above the age (14 years old)
2. when the user is in the last year before the upgrade (13 years old)

File: spec/lib/tasks/kids_spec.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  let(:minor_to_promote) { create(:user, :confirmed, organization:) }
+  let!(:minor_to_promote_account) { create :minor_account, tutor: user, minor: minor_to_promote }
+  let!(:minor_data_to_promote) { create :minor_data, user: minor_to_promote, birthday: }
+  let(:birthday) { 15.years.ago }
+  let(:enable_minors_participation) { false }
+  let!(:minors_organization_config) { create(:minors_organization_config, organization:, enable_minors_participation:) }
+
+  before do
+    minor.confirm
+  end
```
Comment (microstudi) [inline spec/lib/tasks/kids_spec.rb]:
```suggestion
```
This should be unnecessary

**Review by microstudi — CHANGES_REQUESTED** [review]
Thanks @apuertadelaguila 
I've left you a final set of comments. The most important thing is to test more cases for when the upgrade should or not happen:

- 12 years old: not to happen
- 13 years old: not to happen
- 14 years old: to happen
- 15 years old: to happen

With this, we will be done

File: spec/lib/tasks/kids_spec.rb (hunk 10 above)
Comment (apuertadelaguila) [reply spec/lib/tasks/kids_spec.rb]: Done

File: spec/lib/tasks/kids_spec.rb (hunk 12 above)
Comment (apuertadelaguila) [reply spec/lib/tasks/kids_spec.rb]: Done

File: spec/lib/tasks/kids_spec.rb (hunk 11 above)
Comment (apuertadelaguila) [reply spec/lib/tasks/kids_spec.rb]: Done

File: spec/lib/tasks/kids_spec.rb (hunk 9 above)
Comment (apuertadelaguila) [reply spec/lib/tasks/kids_spec.rb]: Done

**Review by microstudi — APPROVED** [review]
ok!

**Comment by NilHomedes** [comment]
I can't review this AC:

> - [ ] As a registered minor when I reach the legal age of consent, I receive an email stating that my account has been promoted to adult.

I had tried to create a minor account that receives the email the next day, but in the letter opener I didn't see it. 

Today I just create a new minor account, that will reach the maximum age tomorrow. Could anyone please review if this email is sent?

**Pull request merged**